provider/spark.go

```diff
@@ -868,12 +928,79 @@ func (e *EMRExecutor) RunSparkJob(args []string, store SparkFileStore) error {
 		StepId:    aws.String(stepId),
 	}, waitDuration)
 	if err != nil {
+		errorMessage, err := e.getStepErrorMessage(e.clusterName, stepId)
+		if err != nil {
+			return fmt.Errorf("could not get error message from EMR step: %v", err)
+		}
+
+		if errorMessage != "" {
+			return fmt.Errorf("the EMR step '%s' failed: %s", stepId, errorMessage)
+		}
+
 		e.logger.Errorw("Failure waiting for completion of EMR cluster", err)
-		return err
+		return fmt.Errorf("failure waiting for completion of EMR cluster: %s", err)
 	}
 	return nil
 }
 
+func (e *EMRExecutor) getStepErrorMessage(clusterId string, stepId string) (string, error) {
+	stepResults, err := e.client.DescribeStep(context.TODO(), &emr.DescribeStepInput{
+		ClusterId: aws.String(clusterId),
+		StepId:    aws.String(stepId),
+	})
+	if err != nil {
+		return "", fmt.Errorf("could not get information on EMR step '%s'", stepId)
+	}
+
+	if stepResults.Step.Status.State == "FAILED" {
+		var errorMsg string
+		// check if there are any errors
+		if stepResults.Step.Status.FailureDetails.Message != nil {
+			// get the error message
+			errorMsg = *stepResults.Step.Status.FailureDetails.Message
+		}
+		if errorMsg != "" {
+			return errorMsg, nil
+		}
+
+		if stepResults.Step.Status.FailureDetails.LogFile != nil {
+			logFile := *stepResults.Step.Status.FailureDetails.LogFile
+
+			bucket, path := getBucketAndPathFromFilePath(logFile)
+			outputFilePath := fmt.Sprintf("%s/stdout.gz", path)
+
+			// wait until log file exists
+			for {
+				fileExists, err := e.logFileStore.Exists(outputFilePath)
+				if err != nil {
+					return "", fmt.Errorf("could not determine if file '%s' exists: %v", outputFilePath, err)
+				}
+
+				if fileExists {
+					break
+				}
+
+				time.Sleep(2 * time.Second)
```

Comment:
I was mistaken. It can take a few minutes to make the log available. 
```
Because log files are uploaded to Amazon S3 every 5 minutes, it can take a few minutes for the log file uploads to complete after the step completes.
```
from [AWS Documentation on Logs](https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-manage-view-web-log-files.html#emr-manage-view-web-log-files-debug)